test/functional/provider/perl_spec.lua

```diff
@@ -0,0 +1,73 @@
+local helpers = require('test.functional.helpers')(after_each)
+local eq, clear = helpers.eq, helpers.clear
+local missing_provider = helpers.missing_provider
+local command = helpers.command
+local write_file = helpers.write_file
+local eval = helpers.eval
+local retry = helpers.retry
+
+do
+  clear()
+  local reason = missing_provider('perl')
+  if reason then
+    pending(string.format("Missing perl host, or perl version is too old (%s)", reason), function() end)
+    return
+  end
+end
+
+before_each(function()
+  clear()
+end)
+
+describe('perl host', function()
+  teardown(function ()
+    os.remove('Xtest-perl-hello.pl')
+    os.remove('Xtest-perl-hello-plugin.pl')
+  end)
+
+  it('works', function()
+    local fname = 'Xtest-perl-hello.pl'
+    write_file(fname, [[
+      package main;
+      use v5.22.1;
+      use Neovim::Ext;
+      use Neovim::Ext::MsgPack::RPC;
+
+      my $session = Neovim::Ext::MsgPack::RPC::socket_session($ENV{NVIM_LISTEN_ADDRESS});
+      my $nvim = Neovim::Ext::from_session($session);
+      $nvim->command('let g:job_out = "hello"');
+      1;
+    ]])
+    command('let g:job_id = jobstart(["perl", "'..fname..'"])')
+    retry(nil, 3000, function() eq('hello', eval('g:job_out')) end)
+  end)
+
+  it('plugin works', function()
+    local fname = 'Xtest-perl-hello-plugin.pl'
+    write_file(fname, [[
+      package TestPlugin;
+      use parent qw(Neovim::Ext::Plugin);
+
+      @{TestPlugin::commands} = ();
+      @{TestPlugin::specs} = ();
+      sub test_command :nvim_command('TestCommand')
+      {
+        my ($this) = @_;
+        $this->nvim->command('let g:job_out = "hello-plugin"');
+      }
+
+      package main;
+      use v5.22.1;
+      use Neovim::Ext;
+      use Neovim::Ext::MsgPack::RPC;
+
+      my $session = Neovim::Ext::MsgPack::RPC::socket_session($ENV{NVIM_LISTEN_ADDRESS});
+      my $nvim = Neovim::Ext::from_session($session);
+      my $plugin = Neovim::Ext::Plugin::new('TestPlugin', $nvim);
+      $plugin->test_command();
```

Comment:
This should run `TestCommand` but it needs the Host class to start the event loop. I don't know how that works.

Reply:
A plugin should derive from ```Neovim::Ext::Plugin```

https://github.com/jacquesg/p5-Neovim-Ext/blob/master/t/rplugin/perl/TestPlugin.pm

To start the host could be as simple as [start_host()](https://github.com/jacquesg/p5-Neovim-Ext/blob/master/lib/Neovim/Ext.pm#L59)

Reply:
Did I not do that for this test? I used `package` twice, one for the plugin and one for the session. I tried `start_host` before but it assumes `rplugin/perl` in `rtp`. I'm not testing that yet.

Reply:
The syntax to instaniate the plugin isn't quite right. It should eithe be:
```my $plugin = TestPlugin->new ($nvim)```
or
```my $plugin = new TestPlugin ($nvim)```;

https://github.com/jacquesg/p5-Neovim-Ext/blob/master/lib/Neovim/Ext/Plugin.pm#L202

It should also register the plugin like so:
https://github.com/jacquesg/p5-Neovim-Ext/blob/master/t/rplugin/perl/TestPlugin.pm#L7